### PR TITLE
Save Linear IR before emit

### PR DESCRIPTION
**Pull request by gretay-js**
This PR adds command line option to marshal the compiler's intermediate representation of each compilation unit to a separate file. Currently, it is only implemented for saving the Linear IR.

This provides a convenient way for an external tool to get a hold of an intermediate representation and manipulate it using compiler-libs. It is analogous to the way ppx tools handle parse trees, but lets us access IRs down the compilation pipeline. There is no alternative way of doing it, as far as I can see, since the removal of compiler hooks in https://github.com/ocaml/ocaml/pull/2276

Adding this capability has previously been discussed, for example: https://github.com/ocaml/ocaml/pull/1945#issuecomment-410240495, https://github.com/ocaml/ocaml/pull/1945#issuecomment-410247545

Following the conventions established for other compilation articfacts, the IR will be saved to a file named <output_prefix>.cmir-linear. The format is defined in linear_format.ml.

The option is named "-save-ir-after <pass>" and supports only "scheduling" pass in this PR.

I have another patch that supports "-save-ir-after parsing" and "-save-ir-after typing" by invoking existing functionality in the compiler.  It seems redundant with other command line options and I don't have a way to test it, but happy to add it on for completeness if needed.  The passes "parsing" and "typing" were already defined in Clflags.Compiler_passes for use with command line option "-stop-after <pass>".

This PR is on top of PR #8938 in which I propose to add "scheduling" to the existing "-stop-after" option. The only new commit here to review is in https://github.com/ocaml/ocaml/commit/8ea2eb9a96ad06639586099bf814c25e8aa67954

### CODE REVIEW

**Comment by gretay-js** [comment]
ping

File: asmcomp/asmgen.ml
Comment (mshinwell) [inline asmcomp/asmgen.ml]: I would format this with a new line after the `=`, so the record isn't so far across.

File: asmcomp/asmgen.ml
Comment (mshinwell) [inline asmcomp/asmgen.ml]: I would advise `begin` and `end` here and below.

File: asmcomp/asmgen.ml
Comment (mshinwell) [inline asmcomp/asmgen.ml]: I think this conditional should definitely have `begin` and `end`.

File: asmcomp/asmgen.mli
Comment (mshinwell) [inline asmcomp/asmgen.mli]: We have labelled arguments now, so I would add a label, and remove the comment.

File: asmcomp/cmm.ml
Comment (mshinwell) [inline asmcomp/cmm.ml:116]: `set_label_counter` might be more descriptive.  Also, the `assert` should be `Misc.fatal_error`, since the property being checked is not a local property of, say, this file.

File: driver/compenv.ml
Comment (mshinwell) [inline driver/compenv.ml:440]: Again, break this line after the `=`.

File: driver/compenv.ml
Comment (mshinwell) [inline driver/compenv.ml:441]: I think you could make `filter` an optional argument, which will remove several hunks from this diff.

File: driver/main_args.ml
Comment (mshinwell) [inline driver/main_args.ml:110]: Same again about line breaks.

File: otherlibs/unix/.depend
Comment (mshinwell) [inline otherlibs/unix/.depend]: Let's revert the changes to this file, they shouldn't be needed.

File: testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml
Comment (mshinwell) [inline testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml:5]: Why does this have "`-S`"?

File: testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml
Comment (mshinwell) [inline testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml:8]: I'm surprised we need a shell script to check the existence of the `.cmir` files, but I'm not very familiar with `ocamltest`.  I suggest asking @shindere if he could review these few parts of the diff.

File: utils/clflags.ml
Comment (mshinwell) [inline utils/clflags.ml:420]: Add newline here before "`let extension`".

File: file_formats/linear_format.mli
Comment (mshinwell) [inline file_formats/linear_format.mli:20]: "Compiler can optionally save Linear representation of a compilation unit" -> "The compiler can optionally save the Linear representation of a compilation unit"

File: file_formats/linear_format.mli
Comment (mshinwell) [inline file_formats/linear_format.mli:18]: This comment and those subsequent can be doc comments ("`(**`"), I think.

File: file_formats/linear_format.ml
Comment (mshinwell) [inline file_formats/linear_format.ml]: I don't think this comment needs to be here; the `.mli` is descriptive enough.

File: file_formats/linear_format.ml
Comment (mshinwell) [inline file_formats/linear_format.ml]: "Saved because linear and emit depend on cmm.label" -> "Saved because Linearize and Emit depend on Cmm.label"

File: file_formats/linear_format.ml
Comment (mshinwell) [inline file_formats/linear_format.ml]: Here and below, spaces before parentheses to be consistent.

File: file_formats/linear_format.ml
Comment (mshinwell) [inline file_formats/linear_format.ml]: Use `String.equal` rather than polymorphic equality.

File: file_formats/linear_format.ml
Comment (mshinwell) [inline file_formats/linear_format.ml]: Redundant parentheses.

File: file_formats/linear_format.mli
Comment (mshinwell) [inline file_formats/linear_format.mli]: "unmashal" -> "unmarshal".

I think a bit more detail could be provided in the comment, in particular that `restore` resets the `Cmm` state and label counter.

**Review by mshinwell — CHANGES_REQUESTED** [review]
This looks fine to me modulo various minor comments.  However since I've been working with @gretay-js on this project, another reviewer with commit access should please have a brief look through the code, and confirm that the approach is reasonable.

**Comment by shindere** [comment]
Mark Shinwell (2019/11/26 05:17 -0800):
> I'm surprised we need a shell script to check the existence of the
> `.cmir` files, but I'm not very familiar with `ocamltest`.  I suggest
> asking @shindere if he could review these few parts of the diff.

I confirm that the shell script is indeed needed, yes.

One thing, though:

I don't think `ocamlopt_byte_exit_status = "0"` is required because
that's what is expected by default.

As far as tests are ocncerned the rest looks okay to me, meaning it's
done as nicely as ocamltest permits. :)

File: asmcomp/cmm.ml
Comment (gretay-js) [reply asmcomp/cmm.ml:116]: The function name `set_label`  goes along with existing `new_label` and `cur_label` all of which set label_counter. The name `label_counter` is not exposed at all in the interface.  I changed the type of `set_label` from `int` to `label` to make it more descriptive and better aligned with the other two functions. Is it enough or do you think it would be better to rename this function anyway?

**Comment by gretay-js** [comment]
> I don't think `ocamlopt_byte_exit_status = "0"` is required because that's what is expected by default.

fixed, thanks!

**Comment by gretay-js** [comment]
@mshinwell  @shindere Thank you for the reviews! I addressed all review comments so far in 1200de97288b.

File: asmcomp/cmm.ml
Comment (mshinwell) [reply asmcomp/cmm.ml:116]: This seems ok.

File: testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml
Comment (mshinwell) [reply testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml:5]: (The testsuite changes have been reviewed separately)

File: testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml
Comment (mshinwell) [reply testsuite/tests/tool-ocamlopt-save-ir/save_ir_after_scheduling.ml:8]: (The testsuite changes have been reviewed separately)

File: file_formats/linear_format.mli
Comment (mshinwell) [inline file_formats/linear_format.mli:34]: I would remove "that currently consists of..." -- it will only get out of date.

Comment (mshinwell) [reply file_formats/linear_format.mli:34]: Actually, forget that.

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
The changes are fine, and I'm happy with the whole thing now.  It would be appreciated if someone else can sign off on this as described above, thanks.

**Comment by gretay-js** [comment]
I've rebased it today and the CI is happy. Can this be considered for merging please? It has been waiting for another reviewer to sign it off.

**Review by let-def — APPROVED** [review]
The changes look fine to me too.

**Comment by gretay-js** [comment]
@let-def thank you for the review! I've just rebased the PR.  @gasche can this be merged please, once the CI is finished?

**Comment by gasche** [comment]
I'm happy to merge if I happen to remember to do it in a few hours after the CI has finished. Feel free to ping me if I don't. I added the "merge-me" label, meaning that anyone else will feel encouraged to merge once the CI comes back green.

**Comment by gretay-js** [comment]
Thank you!

**Comment by hhugo** [comment]
@gasche , it's all green ready for you :)

**Pull request merged**
